resize-observer/devicepixel2.html

```diff
@@ -0,0 +1,157 @@
+     <div class="outer"><div></div><div></div><div></div></div>
+     <div class="outer"><div></div><div></div><div></div><div></div></div>
+     ...
+   </div>
+
+It has both #front and #back starting at the top left corner of the page
+so they should both have a devicePixel position of 0,0
+
+It then sets a repeating background-image and background-size on the the
+#back element and all of the .outer children elements so they have a 2x2
+checkerboard image at a size that should draw each pixel of the 2x2 
+checkerboard pattern in devicePixels.
+
+The devicePixelContentBoxSize of all the elements is queried with
+a ResizeObserver.
+
+Finally the background-position of each child element is set so that
+the patterns should align.
+
+In other words
+
+  <div class="outer">
+    <div></div><div></div><div></div>
+  </div>
+
+The first child will be displayed at left = 0 so it background-position is
+set to 0px 0px. The second child should be displayed at 
+
+secondChildLeft = entry(firstChild).devicePixelContentBoxSize[0].inlineSize. 
+
+The 3rd child should be displayed at 
+
+thirdChildLeft = entry(firstChild).devicePixelContentBoxSize[0].inlineSize + 
+                 entry(secondChild).devicePixelContentBoxSize[0].inlineSize
+
+If the element's left edge is an odd value then we need to set the
+background-position of the element's pattern +1 devicePixel.
+
+If what devicePixelContentBox reports is correct the children's patterns
+will align with the pattern in #back and the children will be invisible.
+If what devicePixelContentBox reports is incorrect the patterns will not
+align and it should be clearly visible.
+*/
+
+const frontElem = document.querySelector('#front');
+const backElem = document.querySelector('#back');
+
+const elemToEntry = new Map();
+
+function setPatternInEntries(entries) {
+  for (const entry of entries) {
+    elemToEntry.set(entry.target, {
+      inlineSize: entry.devicePixelContentBoxSize[0].inlineSize,
+      blockSize: entry.devicePixelContentBoxSize[0].blockSize,
+    });
+  }
+
+  const setPatternInEntry = function(elem, devicePixelX, devicePixelY) {
+    const entry = elemToEntry.get(elem);
+    const width = entry.inlineSize;
+    const height = entry.blockSize;
+    setPattern(elem, devicePixelX, devicePixelY, width, height);
+    return width;
+  }
+
+  let devicePixelY = 0;
+  for (const outerElem of frontElem.children) {
+    console.log(outerElem);
+    let devicePixelX = 0;
+    for (const childElem of outerElem.children) {
+      const width = setPatternInEntry(childElem, devicePixelX, devicePixelY);
+      devicePixelX += width;
+    }
+
+    const outerEntry = elemToEntry.get(outerElem);
+    const outerHeight = outerEntry.blockSize;
+    devicePixelY += outerHeight;
+  }
+
+  setPatternInEntry(backElem, 0, 0);
+}
+
+const observer = new ResizeObserver(setPatternInEntries);
+observer.observe(backElem, {box:"device-pixel-content-box"});
+for (let i = 2; i < 15; ++i) {
```

Comment:
Can you just achieve the same goal with a minimal number of elements declared in HTML rather than built in script? That will be a better reduced testcase, and also make it more clear what layout you're achieving

Reply:
> The flexbox tests this indirectly (because varying number of children varies their locations/size).
[...]
>  I'd prefer to explicitly set location/size.

So, I don't think it's possible to both use this test's same relatively-clean "checkerboard overlaying" strategy (which I like) and to **also** directly/explicitly set the position of any element.  If you directly set an element's position to anything other than the origin, then you can't know for sure whether that element will be painted at an even vs. odd device-pixel-position.  And you have to have that information in order to properly set `background-position` on that element (or its children, etc) to align their checkerboard-image-background with the document's checkerboard-image-background.

(That's kind of the point of this test -- if the cumulative `devicePixelBoxSize` measurements of assumed-to-be-adjacent elements are correct, then we'll be able to correctly predict how to set `background-position` on the various checkerboard elements so that they align properly with the checkerboard behind them and won't stand out).

IMO this setup is pretty reasonable.  I could imagine (but probably wouldn't require) simplifying things by making all the flex containers have only 2 flex-items, both inflexible (`flex: none` with specified `width` and `height`), with a variety of widths on the first items in the flex containers (e.g. ranging from 10.0px, 10.1px, etc, up to 11px), so that the second flex item will end up snapping in different directions between different flex containers.  Then we would set the `background-position` on the second flex item based on its assumed device-pixel x/y coordinates, which we would determine based on the previous flex item's observed size and based on the cumulative sizes of the previous flex containers (as the test currently does).

This would let us use explicit hardcoded sizes, and it would sort-of let us explicitly specify positions as well (via positions being trivially inferrable based on the sizes of prior-siblings).

(We could also achieve a similar setup using floats instead of flexbox, but IMO float layout is hairier than flexbox in simple setups like this, particularly when we're using flexbox pretty trivially as this test does with no complex content-based sizing or alignment.)